File: array/calculation/product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prod = 1
        num_of_zeros = 0
        for v in nums:
            if v == 0:
                num_of_zeros += 1
                continue
            prod = prod * v
        ans = []
        for v in nums:
            if v == 0:
                if num_of_zeros >1:
                    ans.append(0)
                elif num_of_zeros  == 1:
                    ans.append(prod)
            else:
                if num_of_zeros >1:
                    ans.append(0)
                elif num_of_zeros  == 1:
                    ans.append(0)
                else:
                    ans.append(prod // v)
        return ans

File: array/calculation/product_of_array_except_self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
    let z_c = 0;
    let prod = 1;

    for (let v of nums) {
        if (v != 0) {
            prod = prod * v
        } else {
            z_c += 1
        }
    }

    let ans = []
    for (let v of nums) {
        if (v == 0) {
            if (z_c > 1) {
                ans.push(0)
            } else if (z_c == 1) {
                ans.push(prod)
            }
        } else {
            if (z_c > 1) {
                ans.push(0)
            } else if (z_c == 1) {
                ans.push(0)
            } else {
                ans.push(prod / v)
            }
        }
    }

    return ans

};
